src/store.c

```diff
@@ -0,0 +1,665 @@
+        if (!list[i]) continue;
+        vma_will_need(list[i]);
+
+        struct vals *ret = vals_merge(vals, list[i]->vals);
+        if (ret) { vals = ret; } else { goto fail_vals; }
+
+        its[it_len].decoder = store_decoder(list[i]);
+
+        cap += list[i]->vma_len;
+        pairs += list[i]->head->keys;
+        it_len++;
+    }
+
+    assert(it_len);
+
+    struct rill_store store = {0};
+    if (!writer_open(&store, file, cap, ts, quant)) {
+        rill_fail("unable to create '%s'", file);
+        goto fail_open;
+    }
+
+    struct indexer *indexer = indexer_alloc(pairs);
+    struct encoder encoder = writer_begin(&store, vals, indexer);
+
+    for (size_t i = 0; i < it_len; ++i) {
+        if (!(coder_decode(&its[i].decoder, &its[i].kv))) goto fail_coder;
+    }
+
+    struct rill_kv prev = {0};
+
+    while (it_len > 0) {
+        size_t target = 0;
+        for (size_t i = 1; i < it_len; ++i) {
+            if (rill_kv_cmp(&its[i].kv, &its[target].kv) < 0)
+                target = i;
+        }
+
+        struct it *it = &its[target];
+        if (rill_likely(rill_kv_nil(&prev) || rill_kv_cmp(&prev, &it->kv) < 0)) {
+            prev = it->kv;
+            if (!coder_encode(&encoder, &it->kv)) goto fail_coder;
+        }
+
+        if (!coder_decode(&it->decoder, &it->kv)) goto fail_coder;
+        if (rill_unlikely(rill_kv_nil(&it->kv))) {
+            memmove(its + target,
+                    its + target + 1,
+                    (it_len - target - 1) * sizeof(its[0]));
+            it_len--;
+        }
+    }
+
+    store.head->keys = encoder.keys;
+    store.head->pairs = encoder.pairs;
+
+    if (!coder_finish(&encoder)) goto fail_coder;
+    writer_close(&store, indexer, store.head->data_off + coder_off(&encoder));
+
+    for (size_t i = 0; i < list_len; ++i) {
+        if (list[i]) vma_dont_need(list[i]);
+    }
+
+    coder_close(&encoder);
+    indexer_free(indexer);
+    free(vals);
+    return true;
+
+  fail_coder:
+    coder_close(&encoder);
+    writer_close(&store, indexer, 0);
+    indexer_free(indexer);
+  fail_open:
+  fail_vals:
+    free(vals);
+    return false;
+}
+
+
+// -----------------------------------------------------------------------------
+// scan
+// -----------------------------------------------------------------------------
+
+const char * rill_store_file(const struct rill_store *store)
+{
+    return store->file;
+}
+
+rill_ts_t rill_store_ts(const struct rill_store *store)
+{
+    return store->head->ts;
+}
+
+size_t rill_store_quant(const struct rill_store *store)
+{
+    return store->head->quant;
+}
+
+size_t rill_store_vals(const struct rill_store *store)
+{
+    return store->vals->len;
+}
+
+
+struct rill_pairs *rill_store_query_key(
+        struct rill_store *store, rill_key_t key, struct rill_pairs *out)
+{
+    struct rill_pairs *result = out;
+
+    size_t key_idx = 0;
+    uint64_t off = 0;
+    if (!index_find(store->index, key, &key_idx, &off)) return result;
```

Comment:
This should be `goto fail;` instead? I notice that in `rill_query_key()`, you have
```C
result = rill_store_query_key(query->list[i], key, result);
if (!result) return NULL;
```
So if `key` is not found in the index, and `out` is not `NULL`, this function will return a non-`NULL` value and the check will fail.

Reply:
So `index_find` return indicate whether the key was found or not. Note that the key not being present in the store is not an error and instead we just add no values to the result which is returned as is.

Reply:
Ah ... I mistook the `NULL` return for meaning "not found". Never mind then.